feat(webdav): initial implementation for a WebDAV simple auth provider

dschmidt opened the pull request:
As https://github.com/transloadit/uppy/pull/4621 got stale and @Murderlon suggested to implement only the SimpleAuth part of it... here we go.

I intentionally left out any visual configuration for now, I would like to get the actual functionality merged asap and deal with everything else in a followup.

I couldn't make the translations work (at least not the ones in the AuthForm) in the dev server, any hints for that?
Otherwise it works for me (tm)

dschmidt commented:
Seen the type errors ... will fix tomorrow

dschmidt commented:
@Murderlon fixed the type issues, but I'm clueless about the compare_diff test... it looks like new dependencies should be installed by

https://github.com/transloadit/uppy/blob/44a378af9efc522b8f6b0627b7dd51d4af4cfcc5/.github/workflows/e2e.yml#L110-L114

but it doesn't seem to work?

Murderlon commented on packages/@uppy/companion/src/server/provider/webdav/common.js:
```diff
@@ -0,0 +1,125 @@
+    const { protocol } = new URL(url)
+    const HttpAgentClass = getProtectedHttpAgent({ protocol, allowLocalIPs: !allowLocalUrls })
+
+    const { createClient } = await import('webdav')
```
Any reason why the import is done in this function call instead of global?

dschmidt replied:
tbh I don't remember at all. Maybe I copied it from some other provider, maybe I had issues with typescript types, ... will try to make it a regular import

dschmidt replied:
AH, yes! The webdav package is ESM and companion uses CommonJS - hence dynamic async import needs to be used (or Node 20.17 or 22)

Maybe I should add a comment for that

Murderlon replied:
Node.js LTS now has [stable require(esm)](https://github.com/nodejs/node/pull/55217) so in the future when we drop versions we can just import it :)

Murderlon commented on packages/@uppy/companion/src/server/provider/webdav/index.js:
```diff
@@ -0,0 +1,80 @@
+/**
+ * Adapter for WebDAV servers that support simple auth (non-OAuth).
+ */
+class WebdavSimpleAuthProvider extends WebdavProvider {
```
Can we just make it a single class? Seems unnecessary to have a new base class for only one provider.

Murderlon commented on packages/@uppy/webdav/package.json:
```diff
@@ -0,0 +1,35 @@
+    "file uploader",
+    "uppy",
+    "uppy-plugin",
+    "instagram",
```
```suggestion
    "webdav",
```

Murderlon commented on packages/@uppy/webdav/package.json:
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "@uppy/webdav",
+  "description": "Import files from WebDAV into Uppy.",
+  "version": "3.1.1",
```
```suggestion
  "version": "0.1.0",
```

Murderlon commented on packages/@uppy/webdav/src/Webdav.tsx:
```diff
@@ -0,0 +1,117 @@
+const AuthForm = ({ loading, i18n, onAuth }) => {
+  const [webdavUrl, setWebdavUrl] = useState('')
+
+  const onSubmit = useCallback(
+    (e) => {
+      e.preventDefault()
+      onAuth({ webdavUrl: webdavUrl.trim() })
+    },
+    [onAuth, webdavUrl],
+  )
```
We don't need a `useCallback` for this. Note that formatting will probably complain.


```suggestion
  const onSubmit = (event) => {
      event.preventDefault()
      onAuth({ webdavUrl: webdavUrl.trim() })
   }
```

mifi replied:
oops i just realised you removed this. why is it not needed? generally it's considered best practice to useCallback for props given to react components, to prevent re-rendering the whole subtree on every render.

Murderlon commented on packages/@uppy/webdav/src/Webdav.tsx:
```diff
@@ -0,0 +1,117 @@
+
+    this.defaultLocale = locale
+
+    console.log(locale)
```
```suggestion
```

Murderlon reviewed (commented):
Looks good overall! Thanks a lot for picking this up 👌

Murderlon commented:
> I'm clueless about the compare_diff test

Perhaps reverting your changes for the action and resolving https://github.com/transloadit/uppy/pull/5528#discussion_r1865478265 will help

dschmidt commented:
> > I'm clueless about the compare_diff test
> 
> Perhaps reverting your changes for the action and resolving [#5528 (comment)](https://github.com/transloadit/uppy/pull/5528#discussion_r1865478265) will help

I'm afraid I can't easily change that (as commented above) - do you have any other ideas?
Also I'm still struggling with the translations as written in the PR text - any ideas here?

Apart from that everything seems to work and I've addressed all your comments (I think), thanks for the feedback!

Murderlon commented:
> I'm afraid I can't easily change that (as commented above) - do you have any other ideas?
> Also I'm still struggling with the translations as written in the PR text - any ideas here

I can take a look later

Murderlon commented:
I can't make changes because you don't allow edits from contributors. 

Try to add this to `.github/workflows/e2e.yml`

```diff
        run: |
          git checkout FETCH_HEAD -- packages
+     - run: corepack yarn install
      - run: corepack yarn build:js:typeless
```

dschmidt commented:
> I can't make changes because you don't allow edits from contributors.

Maybe that's because I've sent the PR from the owncloud organization... mh :-\

> Try to add this to `.github/workflows/e2e.yml` (line 127)
> 
> ```diff
>         run: |
>           git checkout FETCH_HEAD -- packages
> +     - run: corepack yarn install
>       - run: corepack yarn build:js:typeless
> ```

Still failing :-\

mifi commented on packages/@uppy/webdav/src/Webdav.tsx:
```diff
@@ -0,0 +1,113 @@
+const AuthForm = ({ loading, i18n, onAuth }) => {
+  const [webdavUrl, setWebdavUrl] = useState('')
+
+  const onSubmit = (event) => {
+    event.preventDefault()
+    onAuth({ webdavUrl: webdavUrl.trim() })
+  }
```
usecallback to prevent unnecessary re-renders?
```suggestion
  const onSubmit = useCallback((e) => {
    e.preventDefault()
    onAuth({ webdavUrl: webdavUrl.trim() })
  }, [onAuth, webdavUrl])
```

dschmidt replied:
haha, @Murderlon told me to remove the useCallback we had in place before...

Murderlon replied:
`useCallback`/`useMemo` are overused and should only be leveraged if you _know_ something is slow, not just for the sake of it.

mifi replied:
when coding it's not always obvious that something is going to become slow once it's in production with real world data. IMO it's safer to just add a useCallback than to find out by surprise later that your app is re-rendering the whole React tree on every key stroke in an `<input>` for example.

Murderlon replied:
But this isn't every keystroke, this is a form submit. General advice (from the React team itself) is to not just add it for the sake of it, in fact it's often slower to do the memorization check than what it saves.

Anyway this is such a minor thing in a big PR like this, we better focus on something else.

mifi reviewed (commented):
Nice stuff!
1. Should we have an e2e test for this or would that be very hard to implement? maybe there's a simple mock webdav server implemented in node.js or we can apt install a webdav server and start it up in cypress?
2. should we have docs at https://github.com/transloadit/uppy.io ?
3. ui not pretty but I guess we can merge it as-is for now. at least it works

mifi commented on .github/workflows/e2e.yml:
```diff
@@ -124,6 +124,7 @@ jobs:
           to test')
         run: |
           git checkout FETCH_HEAD -- packages
+      - run: corepack yarn install
```
```suggestion
```

dschmidt commented:
I will address @mifi 's comments asap, not sure what to do with the useCallback ... and I still need help with the translations, any hints?

dschmidt commented on packages/@uppy/companion/src/server/provider/webdav/index.js:
```diff
@@ -0,0 +1,175 @@
+    const HttpAgentClass = getProtectedHttpAgent({ protocol, allowLocalIPs: !allowLocalUrls })
+
+    // dynamic import because Comanion currently uses CommonJS and webdav is shipped as ESM
+    // can be implemented as regular require as soon as Node 20.17 or 22 is required
```
not to myself: oops, there's also a typo "Comanion"

Murderlon commented:
> I still need help with the translations, any hints?

I'm traveling and away from my laptop till 16 December. @mifi would be great if you can help get this over the finish line.

mifi commented:
I don't think you need to translate anything. You've added them English keys, and AFAIK it will fallback to those.

dschmidt commented:
But it doesnt work! The keys are not picked up is what I'm saying

dschmidt commented:
Incorporated your feedback @mifi, now only the translations aren't working yet afaict.

I guess we could and should have docs, but I would tackle that in a follow up after having added a few customization options.
For now I would love to get the feature in to avoid merge conflicts and iterate then

It would be great to have tests but not being familiar with your setup that would be quite some effort again for me and delay all of this further :-(

mifi commented:
> @mifi ping, any help with the translations?

similar issue maybe? #5542

Murderlon commented:
> @mifi ping, any help with the translations?

Sorry for the long wait, I was hoping this would be resolved while I was gone but I'm back now and can take a look.

mifi commented:
Sorry i had to prioritize node-sdk v4 last week
